pkg/declcfg/declcfg.go

```diff
@@ -0,0 +1,226 @@
+			pkg.Icon = &model.Icon{
+				Data:      dPkg.Icon.Base64Data,
+				MediaType: dPkg.Icon.MediaType,
+			}
+		}
+
+		pkg.Channels = map[string]*model.Channel{}
+		for _, ch := range dPkg.Channels {
+			channel := &model.Channel{
+				Package: &pkg,
+				Name:    ch,
+				Bundles: map[string]*model.Bundle{},
+			}
+			if ch == dPkg.DefaultChannel {
+				pkg.DefaultChannel = channel
+			}
+			pkg.Channels[ch] = channel
+		}
+		pkgs[pkg.Name] = &pkg
+	}
+	return pkgs
+}
+
+func populateModelChannels(pkgs model.Model, dBundles []Bundle) error {
+	for _, dBundle := range dBundles {
+		pkg, ok := pkgs[dBundle.Package]
+		if !ok {
+			return fmt.Errorf("unknown package %q for bundle %q", dBundle.Package, dBundle.Name)
+		}
+		for _, bundleChannel := range dBundle.ChannelEntries() {
+			pkgChannel, ok := pkg.Channels[bundleChannel.ChannelName]
+			if !ok {
+				return fmt.Errorf("unknown channel %q for bundle %q", bundleChannel.ChannelName, dBundle.Name)
+			}
+			props, err := dBundle.GetProperties()
+			if err != nil {
+				return fmt.Errorf("get properties for bundle %q: %v", dBundle.Name, err)
+			}
+			pkgChannel.Bundles[dBundle.Name] = &model.Bundle{
+				Package:          pkg,
+				Channel:          pkgChannel,
+				Name:             dBundle.Name,
+				Version:          dBundle.Version,
+				Image:            dBundle.Image,
+				Replaces:         bundleChannel.Replaces,
+				Skips:            dBundle.Skips(),
+				SkipRange:        dBundle.SkipRange(),
+				ProvidedAPIs:     dBundle.ProvidedAPIs(),
+				RequiredAPIs:     dBundle.RequiredAPIs(),
+				Properties:       props,
+				RequiredPackages: dBundle.RequiredPackages(),
+			}
+		}
+	}
+	return nil
+}
+
+func generateChannelUpgradeGraphs(pkgs model.Model) error {
+	for _, pkg := range pkgs {
+		for _, ch := range pkg.Channels {
+			if err := verifyReplaces(*ch); err != nil {
+				return err
+			}
+			chBundles, err := sortBundles(ch.Bundles)
+			if err != nil {
+				return fmt.Errorf("sort package %q, channel %q: %v", pkg.Name, ch.Name, err)
+			}
+			if len(chBundles) > 0 {
+				ch.Head = chBundles[len(chBundles)-1]
+			}
+		}
+	}
+	return nil
+}
+
+func sortBundles(in map[string]*model.Bundle) ([]*model.Bundle, error) {
+	// 1. Find the tail (i.e. the bundle that doesn't replace any others)
+	var out []*model.Bundle
+	for _, b := range in {
+		if b.Replaces == "" {
+			out = append(out, b)
+		}
+	}
+	if len(out) == 0 {
+		return nil, fmt.Errorf("could not find channel tail in upgrade graph")
+	}
+	if len(out) > 1 {
+		return nil, fmt.Errorf("non-linear upgrade graph: multiple tail bundles found")
+	}
+
+	// 2. Build up from the tail
+	for len(out) != len(in) {
+		head := out[len(out)-1]
+		var next *model.Bundle
+		for _, b := range in {
+			if b.Replaces == head.Name {
+				next = b
+			}
+		}
+		if next == nil {
+			return nil, fmt.Errorf("discontinuous or non-linear upgrade graph")
+		}
+		out = append(out, next)
+	}
+	return out, nil
+}
+
+func verifyReplaces(ch model.Channel) error {
```

Comment:
isn't this verification useful for all potential models (not just ones generated from declarative config?)

Reply:
Agreed. Latest commit adds (mostly) comprehensive validation to the model package.